### ISSUE TITLE
Netbox периодически падает при нажатии кнопки Ctrl+R и при обновлении панели - выходит Exception

### ISSUE THREAD

**Issue by programmerj** [michaellukashov/Far-NetBox#210]
1. Netbox периодически падает при нажатии кнопки Ctrl+R - выходит Exception
Соединяюсь по sftp.
Не стабильно, но периодически ( 1 из 10-20 раз) падает Far Manager по нажатию Ctrl+R.
Проявляется на разных серверах.
2. эта же ошибка:
---------------------------
Far.exe - Application Error
---------------------------
The exception unknown software exception (0x40000015) occurred in the application at location 0x03937148.

выходит после редактирования и сохранения файла на удаленном сервере,
т.е. нажимаю F2 далее Esc, файл сохранился на удаленном сервере, пошло обновление папки и выходит вышеуказанный Exception 
3. вот еще выскочило при переходе в верхнюю папку:

---------------------------
Far.exe - Application Error
---------------------------
The exception unknown software exception (0x40000015) occurred in the application at location 0x03827148. 
4. вот Exception при Ctrl+R:

---------------------------
Far.exe - Application Error
---------------------------
The exception unknown software exception (0x40000015) occurred in the application at location 0x03b57148.

**Comment by VictorVG** [comment]
Расшифровка ошибки проста - Windows (Kernel-Mode): {Fatal Application Exit} %hs +

Сбой произошёл по адресу 0x03b57148, а вот что в данный момент в памяти задачи, какая у вас ОС, версия NetBox, Far, какие скрипты и рунтаймы у вас установлены сюань. По крайней мере я угадать это не берусь. Самое большее что я могу - это помочь вам с расшифровкой ошибки. Как я понимаю это ваш баг-репорт - http://bugs.farmanager.com/view.php?id=3203 ?

Давайте попробуем другой вариант сборки (свой скопируйте в иной каталог): [**FarNetBox-2.3.0_Far3_x64.7z**](https://yadi.sk/d/nwyZ1V-vs2w4P) - v2.3.0.436, собрано в MS VC++2010.

**Comment by programmerj** [comment]
Баг репорт мой, но на него никто не реагирует, у меня far по 10-30 раз в день падает, уже замучился.
Сборку попробую.

**Comment by programmerj** [comment]
раньше у меня Netbox не падал, началось это где-то пол года назад, как обновил его, текущая версии Far:
Far Manager, version 3.0 (build 4700) x86 и Far Manager, version 3.0 (build 4700) x64
ОС: Windows 7 и Windows Server 2003 SP2
Неважно какая ОС, Far падает на разных версиях ОС.

**Comment by programmerj** [comment]
5. еще может упасть при выполнении команды shell, пример: cd /home, но тоже не стабильно

**Comment by 2underscores-vic** [comment]
Та же проблема. Win7 x64, Far 3.0 (build 4700).
Часто падает при обновлении панелей с смене директории при работе через SFTP. Чувство, как будто снова вернулся на Win98 - F2 нажимаю после каждого слова.
Кроме того, не всегда показывается актуальное содержимое при операциях с файлами и возврате в посещённую директорию. Приходится нажимать Ctrl+R.

**Comment by 2underscores-vic** [comment]
Может какую-то диагностическую информацию пособирать для выявления проблемы если не у всех воспроизводится?

**Comment by programmerj** [comment]
>> Кроме того, не всегда показывается актуальное содержимое при операциях с файлами и возврате в посещённую директорию. Приходится нажимать Ctrl+R.
согласен, у меня точно такая же ситуация, плагин кеширует некорректно, в более ранних версиях такого не наблюдалсоь

**Comment by VictorVG** [comment]
Пожалуйста - [**FarNetBox-2.3.0_Far3_x86.7z**](https://yadi.sk/d/21g2r5uis2w6A) версия 2.3.0.436, собрано в VC++2010. 

С самим Far - тут я бы порекомендовал его обновить до b4746 ибо за это время в нём устранено много ошибок способных всем кровь портить.

С внутренним кэшированием дерева - вот это я всегда включаю т.к.  зачастую сервера обновляют дерево ранее чем сей кэш проснётся. Я с этим давно, ещё на первых версиях rush встретился и с тех пор на моих системах кэширование сетевых каталогов включается только для NFS, да и то на быстрых каналах. Иначе можно получить мягко скажем странные результаты. Причём как по вине железяк, так и в виде реакции людей, коим после объяснять людям что это они ошиблись, и порой выясняется что они и азов-то не знают, но будучи убеждены в своей правоте никого не хотят слушать.

Усё вместе Far, часть скриптов и плагинов можно тут отыскать [**Far3**](https://yadi.sk/d/oc1fPSIbhvvwT).

С отдельным багом для кэша - давайте сами для себя поймём что мы видим, а там и решение примем как разберёмся. А то заведём, а он не подтвердится и зачем он нам в таком случае?

При сбоях нам нужны дампы, стек вызовов, сообщение об ошибке. Ну, тут нам может помочь Process Hacker, а так как тут поминалась Server 2003, то на ней придётся использовать версию 2.38 [**Process_Hacker-2.38-bin.7z**](https://yadi.sk/d/0cYXG2nlqyFXf), c Вин7 и новее - [**v2.39 Stable**](https://yadi.sk/d/28vbElvsqe2dj) / [**v3.0 Dev**](https://yadi.sk/d/193Gnglm4Ia5D) (документация в архивах! если не прочитать - будет куча ошибок в применении!). Он позволяет нам через контекстное меню процесса получить доступ к его низкоуровневым структурам в том числе и к стеку тредов и просмотру памяти, и т.д. Им можно глянуть с чем мы встретились. У меня есть кое какие мысли на сей счёт, но надо бы проверить.

**Comment by programmerj** [comment]
Баг с кешированием  действительно есть, я каждый день с этим сталкиваюсь по несколько десятков раз.
При Upload-е файла на сервер плагин обновляет папку куда, залил файл. Но показывает предыдущее состояние панели до аплоада, можно даже видео снять чтобы быть убедительнее.

**Comment by VictorVG** [comment]
Я и по вашему описанию смог представить что вы видите, но может кому другому действительно стоит видео показать . Ведь мы все разные, и что одному понятно из описания другому стоит представить в виде видео. Так что думаю не помешает.

**Comment by programmerj** [comment]
вот видео:
https://youtu.be/8ckyp5C_Jns

**Comment by VictorVG** [comment]
А если увеличить тайм-аут ожидания на вкладке Соединение до 90 -120 сек? Сервера FTP обычно отвечают не быстро и дефолтный период 15 сек зашитый сейчас в код слишком мал для них, да и сервера обычно имеют защитный интервал от атаки UDP -флудом - если с запросы на  установление соединения от клиента поступают до ответа на них сервера, то он считает это Flud-атакой и временно блокирует IP клиента , а при повторе атаки IP клиента автоматически блокируется сервером до вмешательства админа. 

Так что я бы поставил тайм-аут FTP/SFTP примерно 90, а при очень загруженных серверах и все 120 секунд - если ответ сервера придёт раньше хорошо, нет - тогда произойдёт реконнект, да и если данные хранятся на удалённом домене через симлинки  это так же увеличивает необходимое время ожидания ответа сервера.

**Comment by programmerj** [comment]
Этот параметр и так был выставлен в 3600 (был подобран опытным путем ранее), так что это не влияет точно.
Кроме того выделенные сервера, на которых проходят эти ошибки все под моим контролем и в параметрах sshd сервера нет защиты от такого типа флуда, блокировка на стороне сервера наступает только через несколько попыток неудачного коннекта.

Здесь идет речь именно об некорректном кешировании на стороне клиента.
Еще раз повторюсь в более ранних версиях плагина не было подобных проблем.

**Comment by programmerj** [comment]
по поводу:
>> С самим Far - тут я бы порекомендовал его обновить до b4746 ибо за это время в нём устранено много ошибок способных всем кровь портить.

где можно взять эту версию?
вот здесь http://farmanager.com/download.php?l=ru
версия 4744
а здесь: https://yadi.sk/d/oc1fPSIbhvvwT
ver 3.0 build 4744 SVN r14346

**Comment by VictorVG** [comment]
Да, это моя опечатка, и в changelog.txt FarUE3 они есть, но это текст, в следующей ревизии устраню (надоело в changelog.txt Upd xxx писать :) ). Реально на ЯД лежит версия 3.0 b4744 r14346.

**Comment by programmerj** [comment]
Тестирую несколько дней последнюю версию, позже отпишусь о результатах

**Comment by 2underscores-vic** [comment]
Имя сбойного приложения: Far.exe, версия: 3.0.4700.0, отметка времени: 0x575beb99
Имя сбойного модуля: NetBox.dll, версия: 2.3.0.436, отметка времени 0x575bf7bb
Код исключения: 0x40000015
Смещение ошибки: 0x0000000000325c86
Идентификатор сбойного процесса: 0x18cc
Время запуска сбойного приложения: 0x01d1e7d4dc9fbf15
Путь сбойного приложения: C:\Program Files\Far Manager\Far.exe
Путь сбойного модуля: C:\Program Files\Far Manager\Plugins\NetBox\NetBox.dll
Код отчета: 7411e4e7-53df-11e6-8bd3-54ee756b65af

**Comment by VictorVG** [comment]
**2underscores-vic**

Тут уже известно. [**Официальная версия с FarManager.com**]((http://www.farmanager.com/download.php?l=ru) собирается MS VC++2015 и это вылезают "оптимизации и усовершенствования" компилятора сделанные Микрософт. :)

**Comment by 2underscores-vic** [comment]
От же твари... А ежели mingw плагин пересобрать, поможет? Или нужно и сам FAR пересобирать?
И можно какие-нибудь ссылки на информацию про "оптимизации и усовершенствования" в MS VC++2015, касающиеся нашей темы?

**Comment by programmerj** [comment]
Имеет ли место быть это: плагин Netbox использует Shared Memory для кеширования директорий.
Если да, то скорее всего проблема в одновременном доступе к этой Shared Memory с разных экземпляров Far-а.

**Comment by VictorVG** [comment]
Я собираю в VC++2010 и по словам **w17**  этот вариант стабильнее, чем собранный в VC++2015. Был эксперимент когда мы пытались спровоцировать крах плагина и это удалось только в сборке VC++2015.

В MinGW с ходу не соберётся - код под VC++ написан, да и если глянуть https://github.com/michaellukashov/Far-NetBox/blob/master/src/NetBox/CMakeLists.txt много интересного вылезет.

> Имеет ли место быть это: плагин Netbox использует Shared Memory для кеширования директорий.

Как-то не смотрел ибо в голову не приходило поскольку у себя я кэширование каталогов давно выключил (сейчас причины не вспомню).

**Comment by 2underscores-vic** [comment]
> у себя я кэширование каталогов давно выключил
Оно отключается в плагине? Как это сделать? В меню конфигурации плагина не нашёл такого

**Comment by programmerj** [comment]
> Оно отключается в плагине? Как это сделать? В меню конфигурации плагина не нашёл такого
в свойствах соединения (F4) наверху есть стрелка вправо в виде треугольника во вкладке "Directories" есть "Cache visited remote directories" и "Cache directory changes"

**Comment by 2underscores-vic** [comment]
Спасибо, нашёл. Попробую поработать без кеширования

**Comment by VictorVG** [comment]
И попробуйте вариант собранный в VC++2010 - если поглядеть что в [**VC++ 2015 Update 3](https://www.visualstudio.com/en-us/news/releasenotes/vs2015-update3-vs) наменяли плохо становится....

**Comment by 2underscores-vic** [comment]
> если поглядеть что в VC++ 2015 Update 3 наменяли плохо становится....

А ещё он отправку [телеметрии](https://www.infoq.com/news/2016/06/visual-cpp-telemetry) включает в бинарник. Но все эти исправления давно уже пора внести в "самобытную" микрософтовскую интерпретацию Стандарта. Ломается, в основном, изначально неправильно написанный код. Ну и, конечно, правильный тоже временами попадает в эту мясорубку...

А где взять бинарник собранный VC++2010? Выше вижу только ссылку на 32-битную сборку.

**Comment by VictorVG** [comment]
Ну, с телеметрией многие поступают просто - отключают на уровне кода. Так сделано в Far3 (в CRT отрублено), Process Hacker - в этих я видел соответствующий код в сорцах, насколько я знаю и другие так же делают.

[**FarNetBox-2.3.0_Far3_x64.7z**](https://yadi.sk/d/nwyZ1V-vs2w4P) - простите, забыл с ходу ссылку дать. :(

У меня сейчас Far3 новый - b4745 r14348 и там же новая UnRAR.dll v5.40.4.2036 (официальной сборки нет, это из unrarsrc-5.4.4.tar.gz собрана в VC++ 2015 Update 3 вызовов телеметрии не обнаружено).

**Comment by programmerj** [comment]
Получается точно виноват блок кеширования в плагине, как его отключил вообще редко падал, может 1-2 раза за несколько дней.
При включенном кешировании падал чаще.
Сборку взял вот эту ночную: Far30b4744.x86.20160727.msi

**Comment by VictorVG** [comment]
Ну, там версия 2.3.0.436 собранная VC++2015, так что вероятно ещё и он свою лепту вносит...

**Comment by programmerj** [comment]
> Ну, там версия 2.3.0.436 собранная VC++2015, так что вероятно ещё и он свою лепту вносит...

это маловероятно, т.к. с включенным кешированием падает постоянно

**Comment by VictorVG** [comment]
Ну, будем надеяться, а я пока Process Hacker пересоберу - были опечатки в VERSION_INFO нового плагина. Устранили в рабочем порядке, пересоберу.

**Comment by 2underscores-vic** [comment]
После отключения кеширования, за два дня ни одного падения. Версию не менял. Протокол, правда, ещё сменил с SFTP на SCP

**Comment by VictorVG** [comment]
Значит в логике управления памяти кэша причина. Там копаться нужно.

**Comment by programmerj** [comment]
однозначно проблема в кешировании, за несколько дней ни одного падения Far-а

**Comment by VictorVG** [comment]
Ясно. Видимо Миша там что изменит, так что я пока его включать не стану.
